src/rgw/driver/rados/rgw_rados.cc

```diff
@@ -6105,7 +6105,7 @@ int RGWRados::Object::Delete::delete_obj(optional_yield y, const DoutPrefixProvi
   const bool need_invalidate = (r == -ECANCELED);
 
   int64_t poolid = ioctx.get_id();
-  if (r >= 0) {
+  if (r >= 0 || r == -ENOENT) {
```

Comment:
ok. we verified above with `state->exists` that the object existed, so getting `ENOENT` here means we raced with another delete op. calling `index_op.complete_del()` instead of `index_op.cancel()` seems like the right thing to do 👍

but i also think we could handle the original `ETIMEDOUT` error better; if rgw can't determine whether or not the delete succeeded, we shouldn't be calling either of `complete_del()` or `cancel()`. instead, we should leave that pending entry in the index so that bucket listing can recover with `check_disk_state()` and `cls_rgw_suggest_changes()`

so something like this?
```c++
  if (r == -ETIMEDOUT) {
    // noop
  } else if (r >= 0 || r == -ENOENT) {
    // complete
  } else {
    // cancel
  }
```

Reply:
already updated